### PR TITLE
Added tail command

### PR DESCRIPTION
**Pull request by rohitshididnyu**
# PR: Add `tail` Command to xv6

## **Summary**
This PR introduces a new **`tail`** user-space utility for the xv6 operating system.  
`tail` prints the last **N** lines from a file or standard input, supporting both **direct file input** and **piped data**.  

---

## **✅ Features**
- **Supported formats**:  
  - `tail` → last 10 lines (default)  
  - `tail <file>`  
  - `tail -n <N>`  
  - `tail -n <N> <file>`  
  - `tail -5 <file>` (shorthand)  
  - `tail +5 <file>` (from line 5 to end)  
- **Argument handling**:  
  - Missing or invalid arguments  
  - Negative or non-numeric values for `-n`  
  - Too many arguments  
  - Unsupported flags  
- **Error handling**:  
  - Nonexistent or unreadable files  
  - Invalid syntax  
- **Standard input**:  
  - Supports data from pipes, e.g., `cat file | tail`  
- **Help option**:  
  - `tail --help` prints usage and exits  
- **Efficient memory usage**:  
  - Uses a fixed-size circular buffer  
  - Handles line overflow and buffer edge cases in xv6  

---

## **🧪 Test Cases**
Executed and verified in QEMU:  
```
tail 
tail -n 5 
tail -n 0 
tail -n              (invalid syntax)
cat  | tail
cat  | tail -n 2
tail -n -3           (invalid N)
tail -m 3            (unsupported flag)
tail -n three        (non-numeric N)
tail -n 5  extra.txt (too many args)
tail –help                    (usage message)
tail -n                        (error message)
tail -n m                      (error message)
tail -n -4                     (error message)
tail +5 
tail -5 
```
✅ All tests passed with correct output or error messages.

---

## **🛠 Implementation Changes**
- Added `user/tail.c` with full implementation  
- Registered `tail` in `Makefile` under `UPROGS`  
- Rebuilt and tested kernel inside QEMU  

---

## **📌 Note**
To include a file from the host system in the xv6 filesystem, update the `Makefile`:
```make
extra = <filename>

### CODE REVIEW

**Comment by Hashmmath** [comment]
Hey, have you tried executing your tail command with files other than README. If yes, then its amazing your code looks good and if not then try to identify where you should change lines in your files and describe about the change in your comment. Thank you.

**Comment by rohitshididnyu** [comment]
> Hey, have you tried executing your tail command with files other than README. If yes, then its amazing your code looks good and if not then try to identify where you should change lines in your files and describe about the change in your comment. Thank you.

Hello @Hashmmath,

I have indeed tested the feature with other files. I understand your concern, as I added the file name in the PR. I believe you may have assumed that I had only tested the feature with that specific file.

To clarify, I will update my PR description to specify <FileName> instead of README.

**Comment by Hashmmath** [comment]
> > Hey, have you tried executing your tail command with files other than README. If yes, then its amazing your code looks good and if not then try to identify where you should change lines in your files and describe about the change in your comment. Thank you.
> 
> Hello @Hashmmath,
> 
> I have indeed tested the feature with other files. I understand your concern, as I added the file name in the PR. I believe you may have assumed that I had only tested the feature with that specific file.
> 
> To clarify, I will update my PR description to specify instead of README.

Yeah I understand, lets just say that you create a file named "hello.txt" in the user folder or wherever you like, then give any content in the hello.txt file and then run "tail -n 5 hello.txt" is it working and showing you the last 5 lines of the file. If not you will get a message saying "tail cannot open hello.txt" why is it happening and in your comment add where do you have to make changes to achieve that. Thank you.

**Comment by rohitshididnyu** [comment]
> > > Hey, have you tried executing your tail command with files other than README. If yes, then its amazing your code looks good and if not then try to identify where you should change lines in your files and describe about the change in your comment. Thank you.
> > 
> > 
> > Hello @Hashmmath,
> > I have indeed tested the feature with other files. I understand your concern, as I added the file name in the PR. I believe you may have assumed that I had only tested the feature with that specific file.
> > To clarify, I will update my PR description to specify instead of README.
> 
> Yeah I understand, lets just say that you create a file named "hello.txt" in the user folder or wherever you like, then give any content in the hello.txt file and then run "tail -n 5 hello.txt" is it working and showing you the last 5 lines of the file. If not you will get a message saying "tail cannot open hello.txt" why is it happening and in your comment add where do you have to make changes to achieve that. Thank you.

Hey Hi, I checked, it works correctly, I have added a Screenshot for clarification
image

**Comment by KiranGhumare4021** [comment]
Hi @rohitshididnyu I'm afraid this code doesn't cover a lot of negative test cases. As per what you have mentioned in the summary, I tried a few commands, precisely,
1) tail -n <filename>  -> Tries to read from stdin
2) tail -n -5 <filename>  -> No error output for this, simply exits the current process without any error message
3) tail -n m <filename>  -> No error output for this, simply exits the current process without any error message
4) tail -n m -> Tries to read from stdin

**Comment by AranyaAryaman** [comment]
> > > > Hey, have you tried executing your tail command with files other than README. If yes, then its amazing your code looks good and if not then try to identify where you should change lines in your files and describe about the change in your comment. Thank you.
> > > 
> > > 
> > > Hello @Hashmmath,
> > > I have indeed tested the feature with other files. I understand your concern, as I added the file name in the PR. I believe you may have assumed that I had only tested the feature with that specific file.
> > > To clarify, I will update my PR description to specify instead of README.
> > 
> > 
> > Yeah I understand, lets just say that you create a file named "hello.txt" in the user folder or wherever you like, then give any content in the hello.txt file and then run "tail -n 5 hello.txt" is it working and showing you the last 5 lines of the file. If not you will get a message saying "tail cannot open hello.txt" why is it happening and in your comment add where do you have to make changes to achieve that. Thank you.
> 
> Hey Hi, I checked, it works correctly, I have added a Screenshot for clarification image

Hey Rohit! You're on the right track. I think what the question from Hashmmath meant was how'd you handle a different file which was created in the host file system (linux or whatever it is). Over here, you're creating a file sample.txt by reading from stdin from the xv6 shell which writes directly to the xv6 file system. Can you think of a way to handle adding new files to the build system itself? To think of a utility of the questio - think what would happen if you were to create an exhaustive test case file which could then be tested against various commands, either programmatically or from the shell?

**Comment by rohitshididnyu** [comment]
> Hi @rohitshididnyu I'm afraid this code doesn't cover a lot of negative test cases. As per what you have mentioned in the summary, I tried a few commands, precisely,
> 
> 1. tail -n   -> Tries to read from stdin
> 2. tail -n -5   -> No error output for this, simply exits the current process without any error message
> 3. tail -n m   -> No error output for this, simply exits the current process without any error message
> 4. tail -n m -> Tries to read from stdin

Hi @KiranGhumare4021, thanks for bringing this to my attention, I have updated the code! Have a look and feel free to update me incase you face any issue.
Thanks

**Comment by rohitshididnyu** [comment]
> > > > > Hey, have you tried executing your tail command with files other than README. If yes, then its amazing your code looks good and if not then try to identify where you should change lines in your files and describe about the change in your comment. Thank you.
> > > > 
> > > > 
> > > > Hello @Hashmmath,
> > > > I have indeed tested the feature with other files. I understand your concern, as I added the file name in the PR. I believe you may have assumed that I had only tested the feature with that specific file.
> > > > To clarify, I will update my PR description to specify instead of README.
> > > 
> > > 
> > > Yeah I understand, lets just say that you create a file named "hello.txt" in the user folder or wherever you like, then give any content in the hello.txt file and then run "tail -n 5 hello.txt" is it working and showing you the last 5 lines of the file. If not you will get a message saying "tail cannot open hello.txt" why is it happening and in your comment add where do you have to make changes to achieve that. Thank you.
> > 
> > 
> > Hey Hi, I checked, it works correctly, I have added a Screenshot for clarification image
> 
> Hey Rohit! You're on the right track. I think what the question from Hashmmath meant was how'd you handle a different file which was created in the host file system (linux or whatever it is). Over here, you're creating a file sample.txt by reading from stdin from the xv6 shell which writes directly to the xv6 file system. Can you think of a way to handle adding new files to the build system itself? To think of a utility of the questio - think what would happen if you were to create an exhaustive test case file which could then be tested against various commands, either programmatically or from the shell?

Thanks @AranyaAryaman for the clarification, I have made the necessary changes!

File: user/tail.c [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+      fd = 0;
+    } else if (argv[1][0] == '+' && is_number(argv[1] + 1)) {
+      start_from_line = atoi(argv[1] + 1);
+      fd = 0;
+    } else {
+      fd = open(argv[1], 0);
+      if (fd < 0) {
+        fprintf(2, "tail: cannot open %s\n", argv[1]);
+        exit(1);
+      }
```
Comment (adithyab-20) [inline user/tail.c]: Great job on the implementation overall! I have a few suggestions: 

When you run `tail -n`, it gives you `tail: cannot open -n` error.  Here it tries to open the argument as a file (since the num of args is 2 but not a number). This is fine since you get an error, but it can be confusing. You could handle this separately and say something like: `tail: option requires an argument -- n`

File: user/tail.c [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  if (n < 0) {
+    fprintf(2, "tail: number of lines cannot be negative: %d\n", n);
+    exit(1);
+  }
+  if (start_from_line == 0) start_from_line = 1;
+
+  // Read all lines into circular buffer
+  while (read(fd, &c, 1) == 1) {
```
Comment (adithyab-20) [inline user/tail.c]: This is correct and it works perfectly but you're reading one byte at a time -- it triggers a syscall per character, which can be slow especially for larger files due to the syscall overhead. Maybe you could consider reading in larger chunks (e.g. 512 bytes) into a temporary buffer and then processing it in memory.

File: user/tail.c [hunk 3]
```diff
@@ -0,0 +1,170 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+#define MAX_LINE 128
+#define MAX_LINES 1024
+
+char buf[MAX_LINES][MAX_LINE];
```
Comment (adithyab-20) [inline user/tail.c]: Nice job on using a circular buffer to handle file lengths of any size. But the 1024 line limit (MAX_LINES) means that for example `tail -1500` or `tail +500` on a file with over 1024 lines won't show the expected output due the lines being overwritten. Consider adding an assumption in your write-up and/or documenting this in print_help (nice addition btw!). Alternatively for +N, you could skip the first N-1 lines by counting newlines and print directly without buffering to handle large files better.

Comment (rohitshididnyu) [reply user/tail.c]: Yup, Updated

File: user/tail.c (hunk 2 above)
Comment (rohitshididnyu) [reply user/tail.c]: Yeah noted, I've made changes, let me know id anything else is required

File: user/tail.c (hunk 1 above)
Comment (rohitshididnyu) [reply user/tail.c]: Thanks!
I have update the error handling
